Update to Libav 12.

jstebbins opened the pull request:
I've rebased the patch set on master to create a complete PR that includes Rodeo's initial porting. This supersedes PR https://github.com/HandBrake/HandBrake/pull/413

twalker314 committed:
libav: remove our custom QSV wrapper.

It prevents us from easily upgrading libav to a newer version.

Note: this breaks builds with --enable-qsv

twalker314 committed:
libhb: add code to support our QSV-accelerated encoder.

The files are copied verbatim from the patch we just removed, where they were known as:
    libavcodec/qsv.h
    libavcodec/qsv.c

twalker314 committed:
libhb: fix build with --enable-qsv

Convert qsv_livav.[ch] to be built within libhb.

Convert other files to use qsv_libav.h in place of libavcodec/qsv.h

Don't attempt to build libav with QSV support (not available in unpatched release/11 branch).

QSV-accelerated decoding is broken, so disable it for the time being.

twalker314 committed:
libhb: don't use av_log.

qsv_libav.[ch] was previosuly use from libavcodec.

twalker314 committed:
qsv: Fix crash attempting to destroy mutex.

Originally part of a2d31ec1c6109424d193832aa83db6958292d797.

twalker314 committed:
contrib: enable avcodec's libmfx (Intel QSV) support when applicable.

twalker314 committed:
decavcodec: enable QSV-accelerated H.264 decoding with libavcodec.

Videos get decoded to NV12 in system memory and converted to our
internal colorspace via libswscale. It should be noted that QSV-
accelerated decoding now works with all our videos encoders :-)

twalker314 committed:
qsv: Update condition under which context is closed.

bradleysepos committed:
contrib: Update to Libav 12.

bradleysepos committed:
contrib: Add libav colr patch to support new color signaling in mov.

Needed for BT.2020, etc.

jstebbins committed:
potential fix for hard to reproduce crash

jstebbins committed:
fix chapter markers when encoding with libav video encoder

encavcodec saved the chapter mark and set the buffers new_chap to 0.
Then work copied the zero'd new_chap over the good mark that encavcodec
set.

jstebbins committed:
decpgssub: fix use of deprecated rect->pict

jstebbins committed:
muxavformat: fix use of deprecated libav interfaces

AVStream.codec is deprecated, use AVStream.codecpar
av_bitstream_filter is deprecated, use av_bsf

jstebbins committed:
stream: fix use of deprecated libav interfaces

AVStream.codec is deprecated, use AVStream.codecpar

jstebbins committed:
decavcodec: fix use of deprecated libav interfaces

avcodec_decode_audio4 is deprecated, use avcodec_send_packet and
avcodec_receive_frame

av_bitstream_filter is deprecated, use av_bsf

jstebbins committed:
encavcodec: fix use of deprecated libav interfaces

AVCodecContext.coded_frame is deprecated.  We didn't really need it.

avcodec_encode_video2 is deprecated, use avcodec_send_frame and
avcodec_receive_packet

jstebbins committed:
encavcodecaudio: fix use of deprecated libav interfaces

avcodec_encode_audio2 is deprecated, use avcodec_send_frame and
avcodec_receive_packet

jstebbins committed:
qsv_libav: silence compiler warning

warning: "/*" within comment
It's very noisy  since this header gets pulled into everything.

jstebbins committed:
qsv_memory: fix compiler warnings

jstebbins committed:
qsv: fix more compiler warnings

jstebbins committed:
qsv: rename av_qsv* to hb_qsv*

bradleysepos commented on libhb/encavcodec.c:
```diff
@@ -530,125 +503,147 @@ static hb_buffer_t * process_delay_list( hb_work_private_t * pv, hb_buffer_t * b
+    // set to a frame number.  Setting it to an actual pts causes
+    // jerky video.
+    // frame->pts = in->s.start;
+    frame.pts = pv->frameno_in++;
```
Still necessary?

jstebbins replied:
This is an ancient issue.  I haven't re-investigated it since originally working around.  I think I'll add it as an issue to remind me to investigate and hopefully fix.

bradleysepos commented:
I read over the changes and didn't see any obvious issues. Nice work.

Does this build for all platforms? Checked for leaks?

jstebbins commented:
I haven't tried building for mingw yet.  But I assume scott has since he's been trying to repro that crash.  I did a batch encode of my samples (329 files) and didn't see memory consumption increases, but I haven't done a valgrind test yet.  I know that there are a couple small leaks in current master that I have a patch for.  So I think I'll commit that patch then rebase this PR and do another valgrind test.

bradleysepos commented:
Sounds good.

galad87 commented:
In my, albeit limited, testing on macOS, seems to be ok.

jstebbins committed:
unref AVPackets, fix merge issues

bradleysepos commented:
Needs another rebase but otherwise, 👍. Builds on all platforms, basic testing performed (already did extensive testing previously).

jstebbins committed:
stream: fix SRT detection

libav changed the codec id they use for SRT in mkv

bradleysepos commented:
Whatever the final commit ends up being, the commit message should be amended with:

Closes #46, closes #276, and closes #416.

jstebbins committed:
fix memory leaks

The series of commits starting with Tim's
"libav: remove our custom QSV wrapper."
Closes #46, closes #276, and closes #416.

jstebbins commented:
Passes my testing now.  It's good to go as far as I'm concerned.

bradleysepos commented:
See my comment about amending the final commit's message to close some issues/PRs, and ship it!
